### DISCUSSION TITLE
Easy change of default connection pool class

### DISCUSSION THREAD

**Discussion by stinovlas** [procrastinate-org/procrastinate#1011] [Ideas]
`PsycopgConnector` currently uses `psycopg_pool.AsyncConnectionPool` by default. It's possible to pass `pool` argument to `App.open_async` which overrides this, but it's not as easy to pass it to procrastinate `worker` / `defer` CLI commands. I ended up with overriding `open_async` method of `PsycopgConnector` like this:

```python
import psycopg_pool
from procrastinate import PsycopgConnector

class NullPoolPsycopgConnector(PsycopgConnector):
    async def open_async(self, pool: psycopg_pool.AsyncConnectionPool | None = None) -> None:
        """Override open_async to use null connection pool by default."""
        if pool is None:
            pool = psycopg_pool.AsyncNullConnectionPool(**self._pool_args)
        return await super().open_async(pool)
```

This works, but it doesn't feel very clean. I'd rather have `connection_pool_cls` kwarg on `PsycopgConnector` where I could pass `AsyncNullConnectionPool` (or any other class inheriting from `AsyncConnectionPool`). If specified, this class would be used instead of `AsyncConnectionPool` in `PsycopgConnector._create_pool`. It would be used like this:

```python
app = TasksApp(
    connector=PsycopgConnector(
        connection_pool_cls=psycopg_pool.AsyncNullConnectionPool,
        json_dumps=orjson.dumps,
        json_loads=orjson.loads,
        **connector_options
    ),
    import_paths=[...],
)
```

What do you think?

**Comment by ewjoachim** [comment]
It makes sense, I like it. I wonder if it should be a "pool class" or a "pool factory" (those are conceptually the same thing: a callable that returns a pool instance), but it could well be any other callable and not a class)

Would you like to contribute this ?

**Reply by stinovlas** [reply]
Pool factory sounds good 🙂. I'll look into it in the upcoming week.

**Reply by ewjoachim** [reply]
https://github.com/procrastinate-org/procrastinate/issues/1022
